### PR TITLE
feat: add `array/base/join`

### LINKED ISSUE #1327: [RFC]: add `@stdlib/array/base/join`
### Description

This RFC proposes adding the package `@stdlib/array/base/join`.

The package should have a structure similar to [`@stdlib/array/base/slice`](https://github.com/stdlib-js/stdlib/tree/64931c5fdb4abc3a339b619658d3b5acc089e556/lib/node_modules/%40stdlib/array/base/slice) and should similarly check for the existence of a `join` method. If not present, the function should manually construct the output string and similarly treat `null` and `undefined` as an empty string (see [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/join#description)).

```javascript
function join( x, separator ) {...}
```

The `separator` argument should be required.

Package: `@stdlib/array/base/join`
Alias: `join`.

### Related Issues

None.

### Questions

No.

### Other

No.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by adityacodes30**
Resolves #1327 .

## Description

> What is the purpose of this pull request?

This pull request adds `array/base/join`.

## Related Issues

> Does this pull request have any related issues?

This pull request:

- closes #1327 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/array/base/join/README.md
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/README.md]:
```suggestion
> Return a string created by joining array elements using a specified separator.
```

File: lib/node_modules/@stdlib/array/base/join/README.md
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/README.md]:
```suggestion
Returns a string created by joining array elements using a specified separator.
```

In general, prefer consistent descriptions.

File: lib/node_modules/@stdlib/array/base/join/README.md
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/README.md]:
```suggestion
-   If provided an array element is either `null` or `undefined`, the function will serialize the element as an empty string.
```

File: lib/node_modules/@stdlib/array/base/join/README.md
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/README.md]:
```suggestion
-   If an array element is either `null` or `undefined`, the function will serialize the element as an empty string.
```

File: lib/node_modules/@stdlib/array/base/join/benchmark/benchmark.length.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/benchmark/benchmark.length.js:65]: You are missing the final assertion check. This is necessary in order to prevent certain compiler optimizations. It is not clear why you did not include this assertion.

Comment (kgryte) [reply lib/node_modules/@stdlib/array/base/join/benchmark/benchmark.length.js:65]: Please see the `@stdlib/bench/harness` README.

File: lib/node_modules/@stdlib/array/base/join/docs/repl.txt
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/docs/repl.txt]: Use the description from the rEADME.

File: lib/node_modules/@stdlib/array/base/join/docs/types/index.d.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/docs/types/index.d.ts]: Use README description.

File: lib/node_modules/@stdlib/array/base/join/docs/types/index.d.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/docs/types/index.d.ts]: The output string is not correct. Also, the function name is not `slice`.

File: lib/node_modules/@stdlib/array/base/join/docs/types/index.d.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/docs/types/index.d.ts]: Use README description.

File: lib/node_modules/@stdlib/array/base/join/docs/types/index.d.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/docs/types/index.d.ts]: This is not correct.

File: lib/node_modules/@stdlib/array/base/join/docs/types/index.d.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/docs/types/index.d.ts]:
```suggestion
* var out = join( x, '-' );
```

File: lib/node_modules/@stdlib/array/base/join/docs/types/test.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/docs/types/test.ts]: The function does not return an array.

File: lib/node_modules/@stdlib/array/base/join/docs/types/test.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/docs/types/test.ts]:
```suggestion
	join( [ 1, 2, 3 ], ',', {} ); // $ExpectError
```

The second argument should be a string; otherwise, this test does not assert what it is supposed to.

File: lib/node_modules/@stdlib/array/base/join/lib/index.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/lib/index.js]: Use README description.

File: lib/node_modules/@stdlib/array/base/join/lib/index.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/lib/index.js:29]:
```suggestion
*
* var x = [ 1, 2, 3, 4 ];
```

Empty line separating imports from expressions.

File: lib/node_modules/@stdlib/array/base/join/lib/index.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/lib/index.js]:
```suggestion
* var out = join( x, ',' );
```

Use single quotes.

File: lib/node_modules/@stdlib/array/base/join/lib/index.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/lib/index.js]:
```suggestion
* // returns '1,2,3,4'
```

File: lib/node_modules/@stdlib/array/base/join/lib/main.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/lib/main.js]: Use modified README description.

Comment (kgryte) [reply lib/node_modules/@stdlib/array/base/join/lib/main.js]: `accessor` is misspelled.

File: lib/node_modules/@stdlib/array/base/join/lib/main.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/lib/main.js]: Don't double call `get`. Cache the value: `v = get( data, i );`.

You don't know how expensive that call is, especially if the target array lazily manifests an array element.

File: lib/node_modules/@stdlib/array/base/join/lib/main.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/lib/main.js]: What happens if `null` or `undefined` is the last element? Here you would append an additional separator, which would be incorrect.

File: lib/node_modules/@stdlib/array/base/join/lib/main.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/lib/main.js]: Same comments as above. Also, cache the value. If `x` is proxied, this can result in significantly degraded performance.

File: lib/node_modules/@stdlib/array/base/join/lib/main.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/lib/main.js]:
```suggestion
	s = '';
```

File: lib/node_modules/@stdlib/array/base/join/lib/main.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/lib/main.js]: Move this to a separate function in the parent scope. This allows JS runtimes to more readily create optimized byte code.

File: lib/node_modules/@stdlib/array/base/join/package.json
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/package.json]:
```suggestion
    "join",
    "concat",
    "concatenate",
    "serialize",
    "tostring"
```

File: lib/node_modules/@stdlib/array/base/join/test/test.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/test/test.js]:
```suggestion
var isString = require( '@stdlib/assert/is-string' );
```

Spaces.

cc @Planeshifter At a minimum, we should create a custom ESLint rule to ensure that all `require` statements always use spaces.

File: lib/node_modules/@stdlib/array/base/join/test/test.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/test/test.js]:
```suggestion
	t.strictEqual( actual, expected, 'returns expected value' );
```

`deepEqual` is inappropriate here. `strictEqual` does value and type equality.

This should be fixed here and elsewhere in this file.

File: lib/node_modules/@stdlib/array/base/join/test/test.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/test/test.js]:
```suggestion
```

This is not applicable.

File: lib/node_modules/@stdlib/array/base/join/test/test.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/test/test.js]:
```suggestion
```

File: lib/node_modules/@stdlib/array/base/join/test/test.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/test/test.js]:
```suggestion
```

File: lib/node_modules/@stdlib/array/base/join/test/test.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/join/test/test.js:39]: You should add tests for edge cases. E.g., how are nulls and undefineds serialized? Complex number arrays? what happens for single element arrays? What about when null or undefined is the last element? What about NaN values? Etc etc. These tests should be replicated for generic, accessor, and typed arrays where appropriate.

**Review by kgryte — CHANGES_REQUESTED** [review]
@adityacodes30 Thank you for working on this. At the moment, this PR needs several changes in order to move forward. 

As some general advice, I would encourage you to not copy-and-paste without carefully considering whether what you are doing is applicable. You copied `array/base/slice`, it seemed, and there are a variety of key differences between that package and this one. As such, you need to be careful and think critically about what needs to be modified.

**Comment by kgryte** [comment]
@adityacodes30 Also, please note how I updated your OP and the PR title. The more you can reduce maintainer noise, the quicker you'll be to receive a review and the more likely that your future PRs will be merged.

**Comment by adityacodes30** [comment]
got it @kgryte, thanks for the very descriptive review - does make things a lot clearer for this pr and going forward

File: lib/node_modules/@stdlib/array/base/join/lib/main.js
Comment (adityacodes30) [reply lib/node_modules/@stdlib/array/base/join/lib/main.js]: makes sense, will keep in mind the costs of  operations

File: lib/node_modules/@stdlib/array/base/join/lib/main.js
Comment (adityacodes30) [reply lib/node_modules/@stdlib/array/base/join/lib/main.js]: the one that manually constructs the string right ?

Comment (kgryte) [reply lib/node_modules/@stdlib/array/base/join/lib/main.js]: Yes, the loop which operates on an indexed array.

File: lib/node_modules/@stdlib/array/base/join/test/test.js
Comment (adityacodes30) [reply lib/node_modules/@stdlib/array/base/join/test/test.js:39]: About NaN values being treated, the current implementation converts it to a NaN string which is also the default behaviour according to mdn docs. Do we want to do it any differently?

Comment (adityacodes30) [reply lib/node_modules/@stdlib/array/base/join/test/test.js:39]: Also , in case of Float64 arrays and like , when their method is invoked (it treats null and undefined as per its own join implementation) , it does not produce expected behaviour treating null or undefined as an empty string as our implementation here 

```
var x = new Float64Array( [ 1.0, 2.0, null, undefined ] );
var s = join( x, ',' );
console.log(s);
> '1,2,0,NaN'
```

how should i handle the writing the testcases here as they do not objectively express `join` , rather direct to the internal methods

Comment (kgryte) [reply lib/node_modules/@stdlib/array/base/join/test/test.js:39]: As mentioned above, test edge cases as appropriate. As `null` and `undefined` have no equivalent in typed arrays, don't add tests for them for `Float64Array` and `Int32Array`. 

For `NaN`, yes, the built-in behavior is correct. Meaning, except for `null` and `undefined`, the serialized value should be equivalent to calling `x[i].toString()`.

BTW: given that `null` and `undefined` checks are unnecessary for typed arrays, so we should add additional branching logic similar to

```javascript
if ( hasMethod( x, 'join' ) {
	// Defer to method...
	return out;
}
if ( obj.accessorProtocol ) {
	// Handle accessor arrays...
	return out;
}
if ( obj.dtype === 'generic' || obj.dtype === null ) {
	// Handle generic indexed arrays and array-like objects with explicit handling of `null` and `undefined`...
	return out;
}
// Handle indexed typed arrays where handling of `null` and `undefined` is not necessary...
return out;
```

**Comment by adityacodes30** [comment]
@kgryte i have made the requested changes, let me know if any other edits are needed

File: lib/node_modules/@stdlib/array/base/join/examples/index.js
Comment (Planeshifter) [inline lib/node_modules/@stdlib/array/base/join/examples/index.js:45]:
```suggestion
```

File: lib/node_modules/@stdlib/array/base/join/lib/main.js
Comment (Planeshifter) [inline lib/node_modules/@stdlib/array/base/join/lib/main.js]:
```suggestion
* @returns {string} joined string
```

File: lib/node_modules/@stdlib/array/base/join/lib/main.js
Comment (Planeshifter) [inline lib/node_modules/@stdlib/array/base/join/lib/main.js]:
```suggestion
* @returns {string} joined string
```

File: lib/node_modules/@stdlib/array/base/join/lib/main.js
Comment (Planeshifter) [inline lib/node_modules/@stdlib/array/base/join/lib/main.js]:
```suggestion
* @returns {string} joined string
```

File: lib/node_modules/@stdlib/array/base/join/examples/index.js
Comment (Planeshifter) [inline lib/node_modules/@stdlib/array/base/join/examples/index.js:53]: Can we add a `// =>` annotation here as well?

File: lib/node_modules/@stdlib/array/base/join/examples/index.js
Comment (Planeshifter) [inline lib/node_modules/@stdlib/array/base/join/examples/index.js:59]: Can we add a `// =>` annotation here as well?

File: lib/node_modules/@stdlib/array/base/join/README.md
Comment (Planeshifter) [inline lib/node_modules/@stdlib/array/base/join/README.md:91]: It is project convention that the examples in the `examples/index.js` file match what is inside the code block of the `## Examples` section. Would you mind replacing this code block with the contents of the `examples/index.js` file, please? (note: we do use `var join = require( '@stdlib/array/base/join' );` instead of the relative require path for the package that is used in `examples/index.js`)

**Review by Planeshifter — CHANGES_REQUESTED** [review]
Left a few comments, but after these are addressed it should be ready to land, I think.

File: lib/node_modules/@stdlib/array/base/join/examples/index.js
Comment (adityacodes30) [reply lib/node_modules/@stdlib/array/base/join/examples/index.js:53]: most certainly

**Review by Planeshifter — APPROVED** [review]
LGTM! Will merge once CI has passed.

Thanks for your contributions!

**Pull request merged**
